### PR TITLE
Introduce sequence-number-based recovery

### PR DESCRIPTION
**Pull request by jasontedor**
This commit introduces sequence-number-based recovery. When a replica has fallen out of sync, rather than performing a file-based recovery we first attempt to replay operations since the last local checkpoint on the replica. To do this, at the start of recovery the replica tells the primary what its local checkpoint is. The primary will then wait for all operations between that local checkpoint and the current maximum sequence number to complete; this is to ensure that there are no gaps in the operations that will be replayed from the primary to the replica. This is a best-effort attempt as we currently have no guarantees on the primary that these operations will be available; if we are not able to replay all operations in the desired range, we just fallback to file-based recovery. Later work will strengthen the guarantees.

Relates #10708

### CODE REVIEW

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractRefCounted.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/util/concurrent/AbstractRefCounted.java]: Looks like a typo here

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 2]
```diff
@@ -379,6 +379,7 @@ void setTook(long took) {
         void freeze() {
             freeze.set(true);
         }
+
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: Same here

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -684,13 +670,20 @@ private IndexResult innerIndex(Index index) throws IOException {
...
+                    location = index.origin() != Operation.Origin.LOCAL_TRANSLOG_RECOVERY ? translog.add(operation) : null;
+                }
             } else {
                 // no version conflict
                 if (index.origin() == Operation.Origin.PRIMARY) {
                     seqNo = seqNoService().generateSeqNo();
                 }
 
-                /**
+                /*
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: ?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This was a Javadoc-style comment inside a method where it has no impact on `javadoc`. While `javac` will treat it as a block comment either way, my IDE formats it as a Javadoc-style comment instead of as a block comment and it annoys me.

**Comment by bleskes** [comment]
w00t. I'll review as soon as I catch up with everything.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -684,13 +670,20 @@ private IndexResult innerIndex(Index index) throws IOException {
             final IndexResult indexResult;
             if (checkVersionConflictResult.isPresent()) {
                 indexResult = checkVersionConflictResult.get();
+                // norelease: this is not correct as this does not force an fsync, and we need to handle failures including replication
+                if (indexResult.hasFailure() || seqNo == SequenceNumbersService.UNASSIGNED_SEQ_NO) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: We discussed another solution for this problem in another channel. The gist was to change the way we deal with version conflicts on replicas. The idea was to try do it as another first and then base this PR on it.

**Review by bleskes — CHANGES_REQUESTED** [review]
This looks great. I left some minor comments and suggestions.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I opened #22626 for this.

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 5]
```diff
@@ -361,7 +361,7 @@ public long getTook() {
 
         void setTranslogLocation(Translog.Location translogLocation) {
             if (freeze.get() == null) {
-                assert failure == null : "failure has to be null to set translog location";
+                assert failure == null || translogLocation == null: "failure has to be null to set translog location";
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: This change will no longer be necessary after #22626.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: I pushed 8b0e5014b1e29976f9dd5acf11422b3d7873ef2f.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I pushed 8b0e5014b1e29976f9dd5acf11422b3d7873ef2f to revert the change in preparation for merging master in after #22626 lands there.

File: core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java [hunk 6]
```diff
@@ -57,4 +57,5 @@ public int totalOperations() {
         }
         return null;
     }
+
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java]: I pushed dac513a3fdcb62272f3c0ccbb148c2bb608d2a4b.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 7]
```diff
@@ -423,78 +459,77 @@ public void finalizeRecovery() {
...
+                request.targetNode());
         }
+
         return totalOperations;
     }
 
+    private Translog.Operation getNextOperationFromSnapshot(final Translog.Snapshot snapshot) {
+        try {
+            return snapshot.next();
+        } catch (final IOException ex) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed 81a1e1c79bdda961e66cadb7298d1d8e908581bb.

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 8]
```diff
@@ -94,6 +112,11 @@ public void readFrom(StreamInput in) throws IOException {
         targetNode = new DiscoveryNode(in);
         metadataSnapshot = new Store.MetadataSnapshot(in);
         primaryRelocation = in.readBoolean();
+        if (in.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            startingSeqNo = in.readZLong();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: `readZLong/writeZLong` is safe?

File: core/src/test/java/org/elasticsearch/index/replication/RecoveryDuringReplicationTests.java [hunk 9]
```diff
@@ -57,11 +61,71 @@ public void testIndexingDuringFileRecovery() throws Exception {
...
+            final boolean flushPrimary = randomBoolean();
+            if (flushPrimary) {
+                shards.flush();
+            }
+
+            final IndexShard recoveredReplica = shards.addReplica(originalReplica);
+            shards.recoverReplica(recoveredReplica);
+            if (flushPrimary && replicaHasDocsSinceLastFlushedCheckpoint) {
+                // replica has something to catch up with, but since we flushed the primary, we should fall back to full recovery
+                assertThat(recoveredReplica.recoveryState().getIndex().fileDetails(), not(empty()));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/replication/RecoveryDuringReplicationTests.java]: I pushed 8960522775d1415f4828dee34d934ef9c1f6330c; please review this one carefully. :wink:

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java (hunk 8 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: it is safe, but I think we should standarize on a single way. We opted for writeLong in other places, so that's why I suggested it here. I'm fine with writeZLong but then we should go to the translog and such and change it there too.

File: core/src/main/java/org/elasticsearch/index/seqno/SequenceNumbersService.java [hunk 10]
```diff
@@ -62,8 +62,8 @@ public SequenceNumbersService(
         final long localCheckpoint,
         final long globalCheckpoint) {
         super(shardId, indexSettings);
-        localCheckpointService = new LocalCheckpointService(shardId, indexSettings, maxSeqNo, localCheckpoint);
-        globalCheckpointService = new GlobalCheckpointService(shardId, indexSettings, globalCheckpoint);
+        localCheckpointTracker = new LocalCheckpointTracker(indexSettings, maxSeqNo, localCheckpoint);
+        globalCheckpointTracker = new GlobalCheckpointTracker(indexSettings, globalCheckpoint, logger);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/SequenceNumbersService.java]: I pushed b6e6cc3b7d3c7fdff0e862e984017d705f3b5230.

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 5 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: I integrated master into this branch after #22626 in d71aa16ec9d0dfd0cfb40f8b54ef6f99ea8c771e.

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java (hunk 8 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: Let's standardize on `readZLong/writeZLong`, but let us do that in a separate PR.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: I pushed ab18bde863883549f42cf0f38820f6068bb74045 for now.

File: core/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 5 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/Engine.java]: I also pushed cea70f4d4ac2455c714e1fc8b33e307c16aa55c8.

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 11]
```diff
@@ -310,10 +281,119 @@ public RecoveryResponse newInstance() {
...
+            } else {
+                logger.trace(
+                        "{} preparing for sequence number-based recovery starting at local checkpoint [{}] from [{}]",
+                        recoveryTarget.shardId(),
+                        startingSeqNo,
+                        recoveryTarget.sourceNode());
+            }
+
+            logger.trace("{} preparing shard for peer recovery", recoveryTarget.shardId());
+            recoveryTarget.indexShard().prepareForIndexRecovery();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: I pushed 34fbb378343f0f94e95e9026dd0b2492dc4fe23d.

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 12]
```diff
@@ -169,50 +175,15 @@ private void doRecovery(final long recoveryId) {
...
-                metadataSnapshot = Store.MetadataSnapshot.EMPTY;
-            } catch (Exception e) {
-                // this will be logged as warning later on...
-                logger.trace("unexpected error while listing local files, failing recovery", e);
-                onGoingRecoveries.failRecovery(recoveryTarget.recoveryId(),
-                    new RecoveryFailedException(recoveryTarget.state(), "failed to list local files", e), true);
-                return;
-            }
+            final Optional<StartRecoveryRequest> maybeRequest = getStartRecoveryRequest(recoveryTarget);
+            if (!maybeRequest.isPresent()) return;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: I'm not sure I agree, having tried to rewrite it, I find it easier to reason about as-is. Let's discuss if you feel strongly.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 13]
```diff
@@ -378,20 +377,7 @@ private static SeqNoStats loadSeqNoStatsFromLuceneAndTranslog(
...
-                assert localCheckpoint == SequenceNumbersService.NO_OPS_PERFORMED;
-                localCheckpoint = Long.parseLong(entry.getValue());
-            } else if (key.equals(MAX_SEQ_NO)) {
-                assert maxSeqNo == SequenceNumbersService.NO_OPS_PERFORMED : localCheckpoint;
-                maxSeqNo = Long.parseLong(entry.getValue());
-            }
-        }
-
-        return new SeqNoStats(maxSeqNo, localCheckpoint, globalCheckpoint);
+        return SequenceNumbers.loadSeqNoStatsFromLuceneCommit(globalCheckpoint, indexWriter.getLiveCommitData());
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I agree; I pushed 1929c03d06ba1b63a2cdd521a8c527e4c1d2abe0 (I think that this will also fit better with future developments).

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 14]
```diff
@@ -169,14 +173,45 @@ public void startPrimary() throws IOException {
...
-            final IndexShard replica = newShard(shardId, false, "s" + replicaId.incrementAndGet(), indexMetaData,
-                () -> { throw new AssertionError("replicas can't sync global checkpoint"); }, null);
+            final IndexShard replica =
+                newShard(shardId, false, "s" + replicaId.incrementAndGet(), indexMetaData, replicaGlobalCheckpointSyncer, null);
             replicas.add(replica);
             updateAllocationIDsOnPrimary();
             return replica;
         }
 
+        public synchronized IndexShard addReplica(IndexShard replica) throws IOException {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: Okay, I pushed c0169c29075a6f315c3467863bce987fbc3fd9cf.

**Comment by jasontedor** [comment]
@bleskes I think this is ready for you again.

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java (hunk 12 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: OK. Here's a patch with what I meant. Talk tomorrow :) 
https://gist.github.com/bleskes/9177f512ebe803dc07dbfdda6f8ef2b7

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks great. Left some more minor feedback.

File: core/src/test/java/org/elasticsearch/index/replication/RecoveryDuringReplicationTests.java [hunk 15]
```diff
@@ -57,11 +61,71 @@ public void testIndexingDuringFileRecovery() throws Exception {
         }
     }
 
+    public void testRecoveryOfDisconnectedReplica() throws Exception {
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/replication/RecoveryDuringReplicationTests.java]: I pushed 5d7d61715a1be2658c63a63fe4898573cebacf58. 😇

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java [hunk 16]
```diff
@@ -94,6 +112,11 @@ public void readFrom(StreamInput in) throws IOException {
         targetNode = new DiscoveryNode(in);
         metadataSnapshot = new Store.MetadataSnapshot(in);
         primaryRelocation = in.readBoolean();
+        if (in.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
+            startingSeqNo = in.readLong();
+        } else {
+            startingSeqNo = SequenceNumbersService.UNASSIGNED_SEQ_NO;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: I pushed 320301c36d79c1b86f9edab3ff83292b8269d059.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 17]
```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
...
     /**
+     * Determines if the source translog is ready for a sequence-number-based peer recovery. The main condition here is that the source
+     * translog contains all operations between the local checkpoint on the target and the current maximum sequence number on the source.
+     *
+     * @param translogView a view of the translog on the source
+     * @return {@code true} if the source is ready for a sequence-number-based recovery
+     * @throws IOException if an I/O exception occurred reading the translog snapshot
+     */
+    boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View translogView) throws IOException {
+        final long startingSeqNo = request.startingSeqNo();
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed 50d31911ec8cd2b0576814ec537d89ad01b67390.

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 18]
```diff
@@ -207,6 +210,11 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
 
     }
 
+    public SeqNoStats loadSeqNoStats(final long globalCheckpoint) throws IOException {
+        final IndexCommit indexCommit = DirectoryReader.listCommits(directory).get(0);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: I pushed 2596dcd547af52f5da8af0e63d0064b448c7538c.

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 19]
```diff
@@ -207,6 +210,11 @@ private static SegmentInfos readSegmentsInfo(IndexCommit commit, Directory direc
 
     }
 
+    public SeqNoStats loadSeqNoStats(final long globalCheckpoint) throws IOException {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: I pushed 2596dcd547af52f5da8af0e63d0064b448c7538c.

File: core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java (hunk 12 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: Okay, that's exactly what I did *except* I was trying to preserve the messages [here](https://github.com/elastic/elasticsearch/blob/ee5f8c452264d47181600d00427b32d08b0d60e2/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L197) and [here](https://github.com/elastic/elasticsearch/blob/ee5f8c452264d47181600d00427b32d08b0d60e2/core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L213), and that it uglier than I preferred. If you're okay dropping those and just getting a generic message, and you appear to be, then I'll just do that already. :smile:

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: hehe. yeah. I think the error in the cause exception should be enough to give us the information we need. No need to be heroic about the extra info.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: I pushed cc2002c00f8bd4240ef8f83db788ded2079846f3.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 20]
```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
...
+             * We need to wait for all operations up to the current max to complete, otherwise we can not guarantee that all
+             * operations in the required range will be available for replaying from the translog of the source.
+             */
+            cancellableThreads.execute(() -> shard.waitForOpsToComplete(endingSeqNo));
+
+            final LocalCheckpointTracker tracker = new LocalCheckpointTracker(shard.indexSettings(), startingSeqNo, startingSeqNo - 1);
+            final Translog.Snapshot snapshot = translogView.snapshot();
+            Translog.Operation operation;
+            while ((operation = snapshot.next()) != null) {
+                tracker.markSeqNoAsCompleted(operation.seqNo());
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed 1e59f84799232018b2e4257c55771e1248d7a113.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 21]
```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
...
+            }
+            return tracker.getCheckpoint() >= endingSeqNo;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Then it's fine?

```
   startingSeqNo == endingSeqNo + 1
=> tracker initialized with maxSeqNo == startingSeqNo, localCheckpoint == startingSeqNo - 1
=> tracker initialized with maxSeqNo == endingSeqNo + 1, localCheckpoint == endingSeqNo + 1 - 1
=> tracker initialized with maxSeqNo == endingSeqNo + 1, localCheckpoint == endingSeqNo
=> tracker localCheckpoint at end of while loop is endingSeqNo
=> condition tracker.getCheckpoint() >= endingSeqNo passes
```

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 22]
```diff
@@ -177,6 +190,41 @@ public RecoveryResponse recoverToTarget() throws IOException {
...
+     *
+     * @param translogView a view of the translog on the source
+     * @return {@code true} if the source is ready for a sequence-number-based recovery
+     * @throws IOException if an I/O exception occurred reading the translog snapshot
+     */
+    boolean isTranslogReadyForSequenceNumberBasedRecovery(final Translog.View translogView) throws IOException {
+        final long startingSeqNo = request.startingSeqNo();
+        final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+        logger.trace("starting: [{}], ending: [{}}", startingSeqNo, endingSeqNo);
+        if (startingSeqNo - 1 <= endingSeqNo) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed 1b7eab7258daf930e7e925ca403056c5e6ded228.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 21 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Put more succinctly: if `startingSeqNo - 1` equals `endingSeqNo` then the local checkpoint on the tracker is initialized to the `endingSeqNo` so the condition and the end of the while loop trivially passes.

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 22 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: > I think we should log a warning if we have startSeqNo > endingSeqNo+1 and also assert it doesn't happen (so we fail in tests)

I do not think this is correct. Imagine a snapshot that restores a previous commit. After the snapshot is restored, a recovery will commence from the restored primary to the replica shards. The replica shards will still have files on disk that could have a commit point after the restored commit point. Their local checkpoint could exceed the maximum sequence number on the restored primary. The replicas will send a start recovery request where starting sequence number on the target is greater than maximum sequence number (plus one) on the source.

**Comment by jasontedor** [comment]
@bleskes It's back to you.

**Comment by jasontedor** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java (hunk 16 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/StartRecoveryRequest.java]: This has to be reverted due to test infrastructure that randomized stream versions. I pushed 6a0b70e60b45298f380149ac415a27ea0dac7e42.

**Comment by jasontedor** [comment]
retest this please

File: core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 22 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: hmm. The snapshot and restore aspect is interesting. I think we need to deal with it separately as it basically violates all assumptions on the local/global checkpoint semantics. I added a note to https://github.com/elastic/elasticsearch/issues/10708 about it.  Potentially just force flush so no doc level recovery will be possible. Can you add a norelease here with a note about how a restore  can cause havoc?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I pushed 3571036a2887b50e8b11b0519f3e3bdcd3f0053a.

File: core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class EvilRecoveryIT extends ESIntegTestCase {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java]: maybe `EvilPeerRecoveryIT`?

File: core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+    @Override
+    protected Settings nodeSettings(final int nodeOrdinal) {
+        final Settings nodeSettings = super.nodeSettings(nodeOrdinal);
+        final int processors = randomIntBetween(1, 4);
+        /*
+         * We have to do this to ensure that there are sufficiently many threads to accept the indexing requests, otherwise operations will
+         * instead be queued and never trip the latch that all operations are inside the engine.
+         */
+        return Settings.builder().put(nodeSettings).put("processors", processors).put("thread_pool.bulk.size", 1 + processors).build();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java]: can chance we can flip it around and use the number of processors in the node as a basic for our tests?

**Review by bleskes — APPROVED** [review]
LGTM, though I would love to hear what the reason is for that failing test you told me about. Also, I think we should invest in making the `EvilRecoveryIT` (which is evil indeed) foldable into the `ESIndexLevelReplicationTestCase`. Messing up with engine level execution will also be needed for when we start dealing with failures. I don't want to slow down this PR though.

File: core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java (hunk 24 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java]: I pushed e801c5b0db55093f70c52635ca1fff00939f1e10.

File: core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java (hunk 23 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/indices/recovery/EvilRecoveryIT.java]: I pushed da94b2803694777cf95339609d4b04d60dc837b9.

**Comment by jasontedor** [comment]
@bleskes I pushed adafa21eab67905490d6fdba8d9ca3d06153ee91 but I'm not happy with it. Can you take a look and see if you can come up with something better. A failing test seed is FE6770A74885D66E (be warned, it takes several minutes, and this is the only seed I know of that makes the test fail).

**Comment by jasontedor** [comment]
@bleskes I pushed 06a3785e15eda4862045066a8cd173048989eeaf.

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

**Comment by jasontedor** [comment]
retest this please

**Comment by bleskes** [comment]
retest this please

**Comment by bleskes** [comment]
retest this please

**Comment by jasontedor** [comment]
Thanks @bleskes. :smile:
